File: Algorithms and Abstract Data Types/PA1/backup/List.java
/*
Andy Huang
ahuang44
PA1
*/

class List{
	private class Node{
		int data;
		Node next;
		Node previous;
		Node(int data){
			this.data=data;
			next=null;
			previous=null;
		}
	}
	private Node front;
	private Node back;
	private Node cursor;
	private int length;
	List(){
		front=back=cursor=null;
		length=0;
	}
	// Returns the number of elements in this List.
	int length(){
		return length;
	}
	// If cursor is defined, returns the index of the cursor element, otherwise returns -1.
	int index(){
		Node n=front;
		int index=0;
		if (cursor==null) {
			return -1;
		}
		while(n!=null){
			if (n.data==cursor.data) {
				return index;
			}
			n=n.next;
			index++;
		}
		return index;
	}
	// Returns front element. Pre: length()>0
	int front(){
		if (length==0) {
			throw new RuntimeException("List Error: front() called on empty List");
		}
		return front.data;
	}
	// Returns back element. Pre: length()>0
	int back(){
		if (length==0) {
			throw new RuntimeException("List Error: back() called on empty List");
		}
		return back.data;
	}
	// Returns cursor element. Pre: length()>0, index()>=0
	int get(){
		if (length==0) {
			throw new RuntimeException("List Error: get() called on empty List");
		}
		if (index()!=-1) {
			return cursor.data;
		}else{
			throw new RuntimeException("List Error: get() called on cursor that is off the list");
		}
	}
	//// Returns true if and only if this List and L are the same integer sequence
	boolean equals(List L){
		boolean isEqual=true;
		Node thisTemp=this.front;
		Node LTemp=L.front;
		if (this.length==L.length) {
			while(isEqual&&thisTemp!=null&&LTemp!=null){
				if (thisTemp.data!=LTemp.data) {
					isEqual=false;
					return false;
				}else{
					thisTemp=thisTemp.next;
					LTemp=LTemp.next;
				}
			}
			return true;
		}else{
			return false;
		}
	}
	//Manipulation procedures
	// Resets this List to its original empty state.
	void clear(){
		front=back=cursor=null;
		length=0;
	}
	// If List is non-empty, places the cursor under the front element, otherwise does nothing.
	void moveFront(){
		if (length!=0) {
			cursor=front;

		}
	}
	// If List is non-empty, places the cursor under the back element, otherwise does nothing.
	void moveBack(){
		if (length!=0) {
			cursor=back;
		}
	}
	// If cursor is defined and not at front, moves cursor one step toward front of this List, if cursor is defined and at front, cursor becomes undefined, if cursor is undefined does nothing.
	void movePrev(){
		if (cursor!=null) {
			if (cursor==front) {
				cursor=null;
			}else{
				cursor=cursor.previous;
			}
		}
	}
	// If cursor is defined and not at back, moves cursor one step toward back of this List, if cursor is defined and at back, cursor becomes undefined, if cursor is undefined does nothing.
	void moveNext(){
		if (cursor!=null) {
			if (cursor==back) {
				cursor=null;
			}else{
				cursor=cursor.next;
			}
		}
	}
	// Insert new element into this List. If List is non-empty, insertion takes place before front element.
	void prepend(int data){
		Node temp=new Node(data);
		if (front==null) {
			front=temp;
			back=temp;
			length++;
		}else{
			front.previous=temp;
			temp.next=front;
			front=temp;
			length++;
		}
	}

	// Insert new element into this List. If List is non-empty, insertion takes place after back element.
	void append(int data){
		Node temp=new Node(data);
		if (back==null) {
			front=temp;
			back=temp;
			length++;
		}else{
			back.next=temp;
			temp.previous=back;
			back=temp;
			length++;
		}
	}
	// Insert new element before cursor. Pre: length()>0, index()>=0
	void insertBefore(int data){
		if (length==0) {
			throw new RuntimeException("List Error: insertBefore() called on empty List");
		}
		if (index()<0) {
			throw new RuntimeException("List Error: insertBefore() called on cursor that is off the list");
		}
		if (cursor==front) {
			prepend(data);
		}else{
			Node temp=new Node(data);
			temp.previous = cursor.previous;
            temp.next = cursor;
            cursor.previous.next = temp;
            cursor.previous = temp;
            length++;
		}
	}
	// Inserts new element after cursor. Pre: length()>0, index()>=0
	void insertAfter(int data){
		if (length==0) {
			throw new RuntimeException("List Error: insertAfter() called on empty List");
		}
		if (index()<0) {
			throw new RuntimeException("List Error: insertAfter() called on cursor that is off the list");
		}
		if (cursor==back) {
			append(data);
		}else{
			Node temp=new Node(data);
			temp.next=cursor.next;
			temp.previous=cursor;
			cursor.next.previous=temp;
			cursor.next=temp;
			length++;
		}
	}
	// Deletes the front element. Pre: length()>0
	void deleteFront(){
		if (length==0) {
			throw new RuntimeException("List Error: deleteFront() called on empty List");
		}
		if (index()==0) {
			cursor=null;
		}
		front=front.next;
		length--;
		
	}

	// Deletes the back element. Pre: length()>0
	void deleteBack(){
		if (length==0) {
			throw new RuntimeException("List Error: deleteBack() called on empty List");
		}
		if (index()==length-1) {
			cursor=null;
		}
		back=back.previous;
		length--;

	}

	// Deletes cursor element, making cursor undefined. Pre: length()>0, index()>=0
	void delete(){
		if (length==0) {
			throw new RuntimeException("List Error: delete() called on empty List");
		}
		if (index()<0) {
			throw new RuntimeException("List Error: delete() called on cursor that is off the list");
		}
		if (cursor==front) {
			deleteFront();
		}else if(cursor==back){
			deleteBack();
		}else{
			cursor.previous.next=cursor.next;
			cursor.next.previous=cursor.previous;
			cursor=null;
			length--;
		}
		
	}

	// Overrides Object's toString method. Returns a String representation of this List consisting of a space separated sequence of integers, with front on left.
	public String toString(){
		String fullList="";
		Node temp=front;
		while(temp!=null){
			fullList+=temp.data+" ";
			temp=temp.next;
		}
		return fullList;
	}
	
	// Returns a new List representing the same integer sequence as this List. The cursor in the new list is undefined, regardless of the state of the cursor in this List. This List is unchanged.
	List copy(){
		List listCopy=new List();
		Node temp=front;
		while (temp!=null) {
			listCopy.append(temp.data);
			temp=temp.next;
		}
		listCopy.cursor=null;
		return listCopy;
	}


}//end list class

File: Algorithms and Abstract Data Types/PA4/backup/List.c
/*
Andy Huang
ahuang44
PA4
List.c 
*/

#include <stdio.h>
#include <stdlib.h>
#include "List.h"

typedef struct NodeObj{
	int data;
	struct NodeObj* next;
	struct NodeObj* previous;

}NodeObj;

typedef NodeObj* Node;

Node newNode(int data){
	Node S = malloc(sizeof(NodeObj));
	S->data = data;
	S->next = NULL;
	S->previous = NULL;
	return(S);

}

void freeNode(Node* pN){
	if(pN!=NULL && *pN!=NULL){
    	free(*pN);
    	*pN=NULL;
   	}
}
// Constructors-Destructors ---------------------------------------------------
typedef struct ListObj{
	Node front;
	Node back;
	Node cursor;
	int length;
}ListObj;

List newList(void){
	List L = malloc(sizeof(ListObj));
	L->front=L->back=L->cursor=NULL;
	L->length=0;
	return(L);
}
void freeList(List* pL){
	if(pL!=NULL && *pL!=NULL){
    	while(length(*pL)!=0){
    		deleteFront(*pL);
    	}
     	free(*pL);
     	*pL=NULL;
   }
}
// Access functions -----------------------------------------------------------
int length(List L){
	return(L->length);
}
int index(List L){
	Node N=L->front;
	int index=0;
	if (L->cursor==NULL)
	{
		return -1;
	}
	while(N!=NULL){
		if (N->data==L->cursor->data)
		{
			return index;
		}
		N=N->next;
		index++;
	}
	return index;
}
int front(List L){
	if (L->length==0)
	{
		fprintf(stderr,"List Error: front() called on empty List");
		exit(EXIT_FAILURE);
	}
	return L->front->data;
}
int back(List L){
	if (L->length==0)
	{
		fprintf(stderr,"List Error: back() called on empty List");
		exit(EXIT_FAILURE);
	}
	return L->back->data;
}
int get(List L){
	if (L->length==0)
	{
		fprintf(stderr,"List Error: get() called on empty List");
		exit(EXIT_FAILURE);
	}
	if (index(L)!=-1)
	{
		return L->cursor->data;
	}else{
		fprintf(stderr,"List Error: back() called on cursor that is off the list");
		exit(EXIT_FAILURE);
	}
}
int equals(List A, List B){
	//1 is true 0 is false
	int returnValue=1;
	Node tempA=A->front;
	Node tempB=B->front;
	if (A->length==B->length)
	{
		while(returnValue==1&&tempA!=NULL&&tempB!=NULL){
			if (tempA->data!=tempB->data)
			{
				returnValue=0;
				return returnValue;
			}else{
				tempA=tempA->next;
				tempB=tempB->next;
			}
		}
		return 1;
	}else{
		return 0;
	}
}
// Manipulation procedures ----------------------------------------------------
void clear(List L){
	freeNode(&L->cursor);
	freeNode(&L->front);
	freeNode(&L->back);
	L->length=0;
}
void moveFront(List L){
	if (L->length!=0)
	{
		L->cursor=L->front;
	}
}
void moveBack(List L){
	if (L->length!=0)
	{
		L->cursor=L->back;
	}
}
void movePrev(List L){
	if (L->cursor!=NULL)
	{
		if (L->cursor==L->front)
		{
			L->cursor=NULL;
		}else{
			L->cursor=L->cursor->previous;
		}
	}
}
void moveNext(List L){
	if (L->cursor!=NULL)
	{
		if (L->cursor==L->back)
		{
			L->cursor=NULL;
		}else{
			L->cursor=L->cursor->next;
		}
	}
}
void prepend(List L, int data){
	Node temp=newNode(data);
	if (L->front==NULL)
	{
		L->front=temp;
		L->back=temp;
		L->length++;
	}else{
		L->front->previous=temp;
		temp->next=L->front;
		L->front=temp;
		L->length++;
	}
	//freeNode(&temp);
}
void append(List L, int data){
	Node temp=newNode(data);
	if (L->back==NULL)
	{
		L->front=temp;
		L->back=temp;
		L->length++;
	}else{
		L->back->next=temp;
		temp->previous=L->back;
		L->back=temp;
		L->length++;
	}
	//freeNode(&temp);
}
void insertBefore(List L, int data){
	if (L->length==0)
	{
		fprintf(stderr,"List Error: insertBefore() called on empty List");
		exit(EXIT_FAILURE);
	}
	if (index(L)<0)
	{
		fprintf(stderr,"List Error: insertBefore() called on cursor that is off the list");
		exit(EXIT_FAILURE);
	}
	if(L->cursor==L->front){
		prepend(L,data);
	}else{
		Node temp=newNode(data);
		temp->previous=L->cursor->previous;
		temp->next=L->cursor;
		L->cursor->previous->next=temp;
		L->cursor->previous=temp;
		L->length++;
		//freeNode(&temp);
	}
}
void insertAfter(List L, int data){
	if (L->length==0)
	{
		fprintf(stderr,"List Error: insertAfter() called on empty List");
		exit(EXIT_FAILURE);
	}
	if (index(L)<0)
	{
		fprintf(stderr,"List Error: insertAfter() called on cursor that is off the list");
		exit(EXIT_FAILURE);
	}
	if(L->cursor==L->back){
		append(L,data);
	}else{
		Node temp=newNode(data);
		temp->next=L->cursor->next;
		temp->previous=L->cursor;
		L->cursor->next->previous=temp;
		L->cursor->next=temp;
		L->length++;
		//freeNode(&temp);
	}
}
void deleteFront(List L){
	if (L->length==0)
	{
		fprintf(stderr,"List Error: deleteFront() called on empty List");
		exit(EXIT_FAILURE);
	}
	if (index(L)==0)
	{
		L->cursor=NULL;
	}
	if (L->length==1)
	{
		freeNode(&L->front);
		L->length--;
	}else{
		L->front=L->front->next;
		freeNode(&L->front->previous);
		L->length--;
	}
	
}
void deleteBack(List L){
	if (L->length==0)
	{
		fprintf(stderr,"List Error: deleteBack() called on empty List");
		exit(EXIT_FAILURE);
	}
	if (index(L)==L->length-1)
	{
		L->cursor=NULL;
	}
	if (L->length==1)
	{
		L->front=L->back=NULL;
		freeNode(&L->front);
		//L->length--;
	}else{
		Node temp=L->back;
		L->back=temp->previous;
		freeNode(&temp);
	}
	
	L->length--;
}
void delete(List L){
	if (L->length==0)
	{
		fprintf(stderr,"List Error: delete() called on empty List");
		exit(EXIT_FAILURE);
	}
	if (index(L)<0)
	{
		fprintf(stderr,"List Error: delete() called on cursor that is off the list");
		exit(EXIT_FAILURE);
	}
	if (L->cursor==L->front)
	{
		deleteFront(L);
	}else if(L->cursor==L->back){
		deleteBack(L);
	}else{
		L->cursor->previous->next=L->cursor->next;
		L->cursor->next->previous=L->cursor->previous;
		freeNode(&L->cursor);
		L->cursor=NULL;
		L->length--;
	}
}
// Other operations -----------------------------------------------------------
void printList(FILE* out, List L){
	Node temp=L->front;
	while(temp!=NULL){
		fprintf(out,"%d ",temp->data);
		temp=temp->next;
	}
}
List copyList(List L){
	List listCopy = newList();
	Node temp=L->front;
	while(temp!=NULL){
		append(listCopy,temp->data);
		temp=temp->next;
	}
	listCopy->cursor=NULL;
	return listCopy;
}